ci: use sidecar versions from the build.env

humblec opened the pull request:
The sidecar images in minikube deployment will be fetched from build.env and used/validated accordingly.

Testing ( with all versions set):
```
validating sidecar controller configuration
```
Testing without a sidecar version set:
```
validating sidecar controller configuration
CSI_RESIZER_VERSION version is empty, make sure build.env has set this sidecar version

```

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

Madhu-1 reviewed (commented):
This is supposed to be overridden from the build.env, if we start updating sidecar is both places it's a double work. We should stick to updating it in only one place.

humblec commented:
> This is supposed to be overridden from the build.env, if we start updating sidecar is both places it's a double work. We should stick to updating it in only one place.

Yeah, however its is possible to use this `scripts/minikube.sh` locally  , where the version update make sense. 

As we keep the versions in the script its indeed better to keep the latest versions in place. Also sidecar updates are not going to happen very frequently.. Mostly once with a kube release.  iow, there is no point to keep very old versions in the script either..

Madhu-1 reviewed (commented):
Do source build.env in minikube.sh and use it? and this avoids duplicate works if we maintain image versios in multiple places

humblec commented:
See sourcing can be used, but my point is simple, if we keep versions, keep latest versions in place, and if we dont want to maintain the versions separately in this file,  there is no point to keep the very old versions out there. Also sourcing make the dependency to build.env , if thats the case we dont even need mention of these versions here in this file.

Madhu-1 commented:
> See sourcing can be used, but my point is simple, if we keep versions, keep latest versions in place, and if we dont want to maintain the versions separately in this file, there is no point to keep the very old versions out there. Also sourcing make the dependency to build.env , if thats the case we dont even need mention of these versions here in this file.

I don't like the idea of updating it in multiple places which is completely useless. if you think it can be removed please remove and add a validation to check the image versions are set.

humblec commented:
So, locally it works, however in CI, the `build.env` path is different... any thoughts  here @ceph/ceph-csi-contributors

Madhu-1 commented:
> So, locally it works, however in CI, the `build.env` path is different... any thoughts here @ceph/ceph-csi-contributors

source build.env like below?

```bash
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
# shellcheck source=build.env
source "${SCRIPT_DIR}/../build.env"
```

humblec commented:
> > So, locally it works, however in CI, the `build.env` path is different... any thoughts here @ceph/ceph-csi-contributors
> 
> source build.env like below?
> 
> ```
> SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
> # shellcheck source=build.env
> source "${SCRIPT_DIR}/../build.env"
> ```

I knew this existed some where but was not able to fetch it .. Thanks.. should be good now. :+1:

nixpanic reviewed (changes requested):
The PR description and commit message suggest that this is for deploying (by Helm or `deploy/` yaml files). However this is for CI only?

Maybe it would be good to validate that the deployment files actually have the same versions of the sidecars as listed in `build.env`. Currently the versions in `build.env` may be different in `charts/` and `deploy/`.

humblec commented:
> The PR description and commit message suggest that this is for deploying (by Helm or `deploy/` yaml files). However this is for CI only?
> 
> Maybe it would be good to validate that the deployment files actually have the same versions of the sidecars as listed in `build.env`. Currently the versions in `build.env` may be different in `charts/` and `deploy/`.

I will keep this PR for CI.. Updated commit and PR description.

humblec committed:
ci: fetch sidecar versions from build.env and use it

The sidecar images in minikube deployment will be fetched from
build.env and used/validated accordingly.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@nixpanic kept this PR for CI only and updated. ptal.. thanks.

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -162,11 +162,21 @@ else
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+source "${SCRIPT_DIR}/../build.env"
+
+sidecars=(CSI_ATTACHER_VERSION CSI_SNAPSHOTTER_VERSION CSI_PROVISIONER_VERSION CSI_RESIZER_VERSION CSI_NODE_DRIVER_REGISTRAR_VERSION)
```
I don't think we need to check the sidecars version by default, this script is used for many tasks. creating minikube cluster. deploying Rook. copying sidecar images. creating pools etc. this check should only be called if someone is using this to copy the sidecar images.

humblec replied:
Done.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -162,11 +162,21 @@ else
+
+sidecars=(CSI_ATTACHER_VERSION CSI_SNAPSHOTTER_VERSION CSI_PROVISIONER_VERSION CSI_RESIZER_VERSION CSI_NODE_DRIVER_REGISTRAR_VERSION)
+for controller in "${sidecars[@]}"; do
+	if [[ "${!controller}" ]]; then
```
For checking empty strings, use `-z`.
No need for listing the versions, that is just clutter IMHO.
This check is not always needed, only for deploying?

humblec replied:
sure, used `-z`

nixpanic commented on scripts/minikube.sh:
```diff
@@ -162,11 +162,21 @@ else
+		echo "${controller}" : "${!controller}"
+	else
+		echo "${controller}" version is empty, make sure build.env has set this sidecar version
+               exit 1
```
indention needs to be fixed

nixpanic commented on scripts/minikube.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash -e
-
+# shellcheck disable=SC1091
```
please only do this where needed, not for the whole script

humblec replied:
Done.

humblec commented:
@nixpanic @Madhu-1 all comments are addressed. ptal .. thanks.

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
+source "${SCRIPT_DIR}/../build.env"
+
+sidecars=(CSI_ATTACHER_VERSION CSI_SNAPSHOTTER_VERSION CSI_PROVISIONER_VERSION CSI_RESIZER_VERSION CSI_NODE_DRIVER_REGISTRAR_VERSION)
+for controller in "${sidecars[@]}"; do
```
use sidecar instead of controller for variable name?

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
+# configure csi sidecar version
+# shellcheck disable=SC1091
+function validate_sidecar() {
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
```
function indentation is wrong compared to other functions in the file

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
 }
 
+# configure csi sidecar version
+# shellcheck disable=SC1091
```
why disable check for the whole function?

humblec commented:
@nixpanic @Madhu-1 comments are addressed. ptal. thanks.

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
@Madhu-1 @nixpanic ptal. thanks.

humblec commented:
@Madhu-1 @nixpanic can you please revisit this PR as you have changes requested here?

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -290,6 +298,8 @@ cephcsi)
     copy_image_to_cluster "${CEPHCSI_IMAGE_REPO}"/cephcsi:"${CSI_IMAGE_VERSION}" "${CEPHCSI_IMAGE_REPO}"/cephcsi:"${CSI_IMAGE_VERSION}"
     ;;
 k8s-sidecar)
+    echo "validating sidecar controller configuration"
```
`echo "validating sidecar controller configuration"` to  `echo "validating sidecar images version"`

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
 }
 
+# configure csi sidecar version
+function validate_sidecar() {
```
comment and function name is not matching rename this to configure_sidecar or update the comment?

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
+# configure csi sidecar version
+function validate_sidecar() {
+    SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+# shellcheck disable=SC1091
```
the comment is not indented is this expected?

Madhu-1 reviewed (commented):
LGTM. left few suggestions

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@Rakshith-R @yati1998 ptal.. thanks

yati1998 reviewed (approved):
LGTM

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@Madhu-1 ptal.. thanks.

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.23

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.23

humblec commented:
@Mergifyio rebase

Rakshith-R commented:
@Mergifyio rebase

humblec commented:
@Mergifyio refresh

humblec commented:
@Mergifyio rebase
